### DISCUSSION TITLE
Enum Arguments/Options with custom click_type are always None

### DISCUSSION THREAD

**Discussion by AlessandroPomponio** [fastapi/typer#777] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the Typer documentation, with the integrated search.
- [X] I already searched in Google "How to X in Typer" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to Typer but to [Click](https://github.com/pallets/click).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
import enum
from typing import Any, Optional

import click
import typer
from typer.testing import CliRunner


def test_custom_click_type_enum():
    class OnlyOneCatEnum(enum.Enum):
        CAT = "cat"

    class NoPluralCatsChoice(click.Choice):
        name = "single_cat"

        def convert(
            self,
            value: Any,
            param: Optional[click.Parameter],
            ctx: Optional[click.Context],
        ) -> Any:
            if value.endswith("s"):
                value = value[:-1]
            return OnlyOneCatEnum(value)

    app = typer.Typer()

    @app.command()
    def custom_click_type_enum(
        ctx: typer.Context,
        animal: OnlyOneCatEnum = typer.Argument(
            None, click_type=NoPluralCatsChoice(choices=["cat"])
        ),
        animal_option: OnlyOneCatEnum = typer.Option(
            None, click_type=NoPluralCatsChoice(choices=["cat"])
        ),
    ):
        assert (
            ctx.params["animal"] == OnlyOneCatEnum.CAT
        ), "The animal param was not a cat!"
        assert (
            ctx.params["animal_option"] == OnlyOneCatEnum.CAT
        ), "The animal_option param was not a cat!"
        assert animal == OnlyOneCatEnum.CAT, "The animal Argument was not a cat!"
        assert (
            animal_option == OnlyOneCatEnum.CAT
        ), "The animal_option Option was not a cat!"

    result = CliRunner().invoke(app, ["cats", "--animal-option", "cats"])
    assert result.exit_code == 0
```

### Description

When I define a custom Click type using Enums and set the `click_type`, i see that:
- `ctx.params[parameter_name]` has the correct value set
- `parameter_name` is instead None

This does not appear to happen with Click, where the relevant piece of code becomes:

```python
    @click.command()
    @click.argument("animal", type=NoPluralCatsChoice(["cats"]))
    @click.option("--animal-option", type=NoPluralCatsChoice(["cats"]))
    @click.pass_context
    def custom_click_type_enum(
        ctx: click.Context,
        animal: NoPluralCatsChoice,
        animal_option: NoPluralCatsChoice,
    ):
        assert animal == OnlyOneCatEnum.CAT, "The animal Argument was not a cat!"
        assert (
            animal_option == OnlyOneCatEnum.CAT
        ), "The animal_option Option was not a cat!"

    result = CliRunner().invoke(
        custom_click_type_enum, ["cats", "--animal-option", "cats"]
    )
    assert result.exit_code == 0
```

### Operating System

macOS

### Typer Version

0.11.0

### Python Version

Python 3.10.12

**Comment by brianmedigate** [comment, marked as answer]
If you return `value` rather than `OnlyOneCatEnum(value)` from `NoPluralCatsChoice.convert()`, then `animal` and `animal_option` will correctly contain the enum instance `OnlyOneCatEnum.CAT` and `ctx.params` will contain the string `cat`, which is the intended behavior I believe.

The issue is that `click.Choice` is not expecting `convert()` to return an instance of an enum. When typer creates a `click.Choice` under the hood, it passes only with the values of the enum (and the enum is only dealt with in typer):
https://github.com/tiangolo/typer/blob/968d81df54997fa0219ca2de2aeacbab54daa003/typer/main.py#L779-L783

**Reply by AlessandroPomponio** [reply]
Thank you very much, Brian! 
When reading the documentation for [custom types on the Click documentation](https://click.palletsprojects.com/en/8.1.x/parameters/#implementing-custom-types), it gave me the impression that I had to return the correct type already, but I see this is not the case with typer.
Thank you once again so much! ^^
